### PR TITLE
Add persistent index skeleton

### LINKED ISSUE #2635: [update] persistent primary index

### PR DESCRIPTION
**Pull request by decster**
## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] others

## Which issues of this PR fixes ：

Fixes #2635 

## Problem Summary(Required) ：

This pr add some code skeleton for persistent primary index, it adds some APIs and basic structure, the detailed implementations will be added in future PRs.

### CODE REVIEW

**Review by chaoyli — DISMISSED** [review]

File: be/src/storage/persistent_index.h [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+    uint64_t bucket() const { return (hash >> 8) & (BucketPerPage - 1); }
+    uint64_t tag() const { return hash & 0xff; }
+};
+
+struct L1Checks {
+    std::vector<uint32_t> key_idxes;
+    std::vector<uint64_t> hashes;
+};
+
+class IndexL0 {
```
Comment (DorianZheng) [inline be/src/storage/persistent_index.h:55]: Is it better to name `IndexL0` as `MutableIndex` and name `IndexL1` as `ImmutableIndex`

File: be/src/storage/persistent_index.h [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+    // |old_values|: return old values if key exist, or set to NullValue if not
+    Status erase(size_t n, const uint8_t* keys, IndexValue* old_values);
+
+private:
+    // index storage directory
+    std::string _path;
+    size_t _key_size = 0;
+    size_t _size = 0;
+    EditVersion _version;
+    std::unique_ptr<IndexL0> _l0;
```
Comment (DorianZheng) [inline be/src/storage/persistent_index.h:187]: Do we need MVCC to realize concurrent of background compaction and foreground `put` and `get`

File: be/src/storage/persistent_index.h (hunk 1 above)
Comment (decster) [reply be/src/storage/persistent_index.h:55]: mutable & immutable do not imply the level of the data structure, just part of it's traits

File: be/src/storage/persistent_index.h (hunk 2 above)
Comment (decster) [reply be/src/storage/persistent_index.h:187]: it takes a more complex design and effort to implement MVCC, so currently, it's not supportted

File: be/src/storage/persistent_index.h (hunk 1 above)
Comment (DorianZheng) [reply be/src/storage/persistent_index.h:55]: The info of level can represented by the owner. Say if you have three level, L1 and L2 shares the same code.

Comment (decster) [reply be/src/storage/persistent_index.h]: OK, changed

**Review by sevev — APPROVED** [review]

**Review by chaoyli — APPROVED** [review]

**Review by Astralidea — APPROVED** [review]

**Pull request merged**